internal/pkg/plugin/module.go

```diff
@@ -0,0 +1,202 @@
+// Copyright (c) 2020, Sylabs Inc. All rights reserved.
+// This software is licensed under a 3-clause BSD license. Please consult the
+// LICENSE.md file distributed with the sources of this project regarding your
+// rights to use or distribute this software.
+
+package plugin
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+
+	"github.com/sylabs/singularity/internal/pkg/buildcfg"
+	"github.com/sylabs/singularity/internal/pkg/sylog"
+)
+
+// SingularitySource represents the symlink name which will
+// point to the Singularity source directory.
+const SingularitySource = "singularity_source"
+
+// Module describes a Go module with its corresponding path and version.
+type Module struct {
+	Path    string
+	Version string
+}
+
+// GoMod describes a parsed go.mod file.
+type GoMod struct {
+	Module  Module
+	Go      string
+	Require []Require
+	Exclude []Module
+	Replace []Replace
+}
+
+// Require describes a require directive in go.mod files.
+type Require struct {
+	Path     string
+	Version  string
+	Indirect bool
+}
+
+// Replace describes a replace directive in go.mod files.
+type Replace struct {
+	Old Module
+	New Module
+}
+
+// GetModules parses the go.mod file found in directory and returns
+// a GoMod instance.
+func GetModules(dir string) (*GoMod, error) {
+	var b bytes.Buffer
+	var e bytes.Buffer
+
+	goMod := filepath.Join(dir, "go.mod")
+
+	if _, err := os.Stat(goMod); err != nil {
+		return nil, fmt.Errorf("while getting information for %s: %s", goMod, err)
+	}
+
+	goPath, err := exec.LookPath("go")
+	if err != nil {
+		return nil, fmt.Errorf("while retrieving go command path: %s", err)
+	}
+
+	cmd := exec.Command(goPath, "mod", "edit", "-json", goMod)
+	cmd.Stdout = &b
+	cmd.Stderr = &e
+
+	if err := cmd.Run(); err != nil {
+		return nil, fmt.Errorf("while reading singularity go.mod: %s\nCommand error:\n%s", err, e.String())
+	}
+
+	modules := new(GoMod)
+
+	if err := json.NewDecoder(&b).Decode(modules); err != nil {
+		return nil, fmt.Errorf("while decoding json data: %s", err)
+	}
+
+	return modules, nil
+}
+
+func getRequire(req Require, array []Require) *Require {
+	for _, a := range array {
+		if a.Path == req.Path {
+			return &a
+		}
+	}
+	return nil
+}
+
+func hasReplace(rep Replace, array []Replace) bool {
+	for _, a := range array {
+		if a.Old.Path == rep.Old.Path && a.Old.Version == rep.Old.Version {
+			return true
+		}
+	}
+	return false
+}
+
+// PrepareGoModules returns a byte array containing a generated go.mod matching
+// Singularity modules in use in order to compile/load the plugin with same version
+// of dependencies.
+func PrepareGoModules(pluginDir string) ([]byte, error) {
```

Comment:
I wonder if this should consider semantic versioning, and error out under certain circumstances. Just to run through a few scenarios, and maybe we can discuss what the _expected_ behaviour would be:

1. Singularity and the plugin both require the **same major version, but Singularity requires an older minor version**. In this case, downgrading the package could cause issues:
 a. Say the plugin developer has chosen to use a newer version because it contains a critical security fix. The version is downgraded, the plugin builds, and now has a vulnerability.
 b. Say the plugin developer has chosen to use a newer version because it contains a new API. The `require` is modified, the plugin fails to build.
1. Singularity and the plugin both require the **same major version, but Singularity requires a newer minor version**. This substitution should be fine, assuming the package sticks to semantic versioning.
1. Singularity and the plugin require **different major versions** of the package. In this case, `plugin compile` should probably outright fail with a clear error about incompatible package versions being required?

Reply:
- 1a, I would say that Singularity still remains vulnerable by using the old version, this is a risk by growing the attack surface with additional code from plugins but Singularity is also a big attack surface in case of vulnerable package.

- 1b, yes there is more chance that build fail, maybe adding some warning about a non backwards compatibility in this case.

- 2, should be OK assuming package respect backwards compatibility.

- 3, in this case yes it may be safer to fail with explicit error if the package didn't use import semantic version path.

Otherwise there is the heavy approach to reduce failure by rewriting dynamically in plugin code the import paths with semantic version path to take the expected versions of package (excluding standard packages of course) and "unshare" them from Singularity, thoughts ?

Reply:
I'm tempted to say that we should explicitly require plugins to use the same major version, and <= minor version (case 2 above.) Otherwise, fail with a clear error message stating the package that is incompatible.

The alternative might end up with "my plugin builds with Singularity version X+1 but not X, something is wrong with Singularity." Or worse, "`plugin compile` re-introduced a vulnerability that I explicitly fixed!". If it's an explicit error, the plugin developer is forced to make the decision of whether to support that version of Singularity or not.

Reply:
Fix applied. Also I'm wondering if we should provide an extra flag to at least disable enforcement on plugin/singularity package minor version during `plugin compile` ? as there is many packages versioned like `v0.0.0-20XXXXX...` which may only differ from few commits with just bug fixes